### PR TITLE
Azure Arc extension

### PR DESCRIPTION
**Pull request by bergeron**
- Adds extensions/arc for managing Azure Arc resources.
- Only the Postgres resource is in this PR.  SQL will come next.
  - Overview, connection strings, and properties pages implemented.
- The UI is built with a ModelViewDashboard.
- Communication with the Arc data controller is done via typescript clients and models generated from the controller's swagger specification.
  - See extensions/arc/src/controller/README.md
- I know GitHub makes large PRs hard to read, so let me know if I can make this easier by splitting it up (e.g. removing the generated clients)

### CODE REVIEW

File: extensions/arc/package.json [hunk 1]
```diff
@@ -0,0 +1,41 @@
+{
+  "name": "arc",
+  "displayName": "Azure Arc",
+  "description": "%arc.description%",
+  "version": "1.0.0",
```
Comment (Charles-Gagnon) [inline extensions/arc/package.json:5]: Let's go with version 0.1.0 until we GA

File: extensions/arc/package.json [hunk 2]
```diff
@@ -0,0 +1,41 @@
+{
+  "name": "arc",
+  "displayName": "Azure Arc",
```
Comment (Charles-Gagnon) [inline extensions/arc/package.json:3]: Let's make this a localized string (given that it's a product name it probably won't be any different but keeps it consistent)

File: extensions/arc/package.json [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+  "description": "%arc.description%",
+  "version": "1.0.0",
+  "publisher": "Microsoft",
+  "preview": true,
+  "license": "https://raw.githubusercontent.com/Microsoft/azuredatastudio/master/LICENSE.txt",
+  "icon": "images/extension.png",
+  "engines": {
+    "vscode": "*",
+    "azdata": "*"
```
Comment (Charles-Gagnon) [inline extensions/arc/package.json:12]: >=1.18.0 since we rely on core changes made in this release

File: extensions/arc/src/constants.ts [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+		};
+	}
+}
+
+export namespace cssStyles {
+	export const text = { 'user-select': 'text', 'cursor': 'text' };
+	export const title = { ...text, 'font-weight': 'bold', 'font-size': '14px' };
+	export const tableHeader = { ...text, 'text-align': 'left', 'border': 'none' };
+	export const tableRow = { ...text, 'border-top': 'solid 1px #ccc', 'border-bottom': 'solid 1px #ccc', 'border-left': 'none', 'border-right': 'none' };
+	export const link = { 'user-select': 'text', 'cursor': 'pointer', 'color': '#0078d4', 'text-decoration': 'underline' };
```
Comment (Charles-Gagnon) [inline extensions/arc/src/constants.ts:97]: You should be able to get rid of this - I made some fixes to the hyperlinks so they're styled correctly by default now

File: extensions/arc/src/controller/auth.ts [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+}
+
+/* Retrieves the current setting for whether to ignore SSL verification errors */
+export function getIgnoreSslVerificationConfigSetting(): boolean {
+	const arcConfigSectionName = 'azureArc';
+	const ignoreSslConfigName = 'ignoreSslVerification';
+
+	try {
+		const config = vscode.workspace.getConfiguration(arcConfigSectionName);
+		return config.get<boolean>(ignoreSslConfigName, true);
```
Comment (Charles-Gagnon) [inline extensions/arc/src/controller/auth.ts:72]: So currently we aren't contributing this config. I believe we'll still want to have this option so make sure you're adding it like we do for BDC : https://github.com/microsoft/azuredatastudio/blob/master/extensions/big-data-cluster/package.json#L144

File: extensions/arc/src/extension.ts [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+import { PostgresDashboard } from './ui/dashboards/postgres/postgresDashboard';
+import { ControllerModel } from './models/controllerModel';
+import { PostgresModel } from './models/postgresModel';
+
+export async function activate(context: vscode.ExtensionContext): Promise<void> {
+	IconPathHelper.setExtensionContext(context);
+
+	vscode.commands.registerCommand('arc.managePostgres', async () => {
+		// Controller information
+		const controllerUrl = 'https://0.0.0.0:30080';
```
Comment (Charles-Gagnon) [inline extensions/arc/src/extension.ts:19]: Don't merge these defaults in - just leave them undefined for now in master (even if that mean it's broken there since we don't have a way to launch this command anyways)

File: extensions/arc/src/models/postgresModel.ts [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+		const ramLimit = this._service.spec.scheduling?.resources?.limits?.['memory'];
+		const cpuRequest = this._service.spec.scheduling?.resources?.requests?.['cpu'];
+		const ramRequest = this._service.spec.scheduling?.resources?.requests?.['memory'];
+		const storage = this._service.spec.storage.volumeSize;
+
+		// Prefer limits if they're provided, otherwise use requests if they're provided
+		let nodeConfiguration = `${nodes} node`;
+		if (nodes > 1) { nodeConfiguration += 's'; }
+		if (cpuLimit) {
+			nodeConfiguration += `, ${this.formatCores(cpuLimit)} vCores`;
```
Comment (Charles-Gagnon) [inline extensions/arc/src/models/postgresModel.ts:117]: These values should be localized

File: extensions/arc/src/models/postgresModel.ts [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+	 * 1 -> 1B
+	 * 1K -> 1KB, 1Ki -> 1KiB
+	 * 1M -> 1MB, 1Mi -> 1MiB
+	 * 1G -> 1GB, 1Gi -> 1GiB
+	 * 1T -> 1TB, 1Ti -> 1TiB
+	 * https://kubernetes.io/docs/concepts/configuration/manage-compute-resources-container/#meaning-of-memory
+	 * @param memory The amount + unit of memory to format e.g. 1K
+	 */
+	private formatMemory(memory: string): string {
+		return memory && !memory.endsWith('B') ? `${memory}B` : memory;
```
Comment (Charles-Gagnon) [inline extensions/arc/src/models/postgresModel.ts:150]: This should be localized

File: extensions/arc/src/ui/components/initializingComponent.ts [hunk 9]
```diff
@@ -0,0 +1,40 @@
...
+	}
+
+	/**
+	 * Runs the specified action when the component is initialized. If already initialized just runs
+	 * the action immediately.
+	 * @param action The action to be ran when the page is initialized
+	 */
+	protected eventuallyRunOnInitialized(action: () => void): void {
+		if (!this._initialized) {
+			this.onInitializedPromise.promise.then(() => action()).catch(error => console.error(`Unexpected error running onInitialized action for BDC Page : ${error}`));
```
Comment (Charles-Gagnon) [inline extensions/arc/src/ui/components/initializingComponent.ts:34]: Update BDC reference here

File: extensions/arc/src/ui/dashboards/postgres/postgresBackupPage.ts [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+	protected get id(): string {
+		return 'postgres-backup';
+	}
+
+	protected get icon(): { dark: string; light: string; } {
+		return IconPathHelper.backup;
+	}
+
+	protected get container(): azdata.Component {
+		return this.modelView.modelBuilder.text().withProperties<azdata.TextComponentProperties>({ value: 'Backup' }).component();
```
Comment (Charles-Gagnon) [inline extensions/arc/src/ui/dashboards/postgres/postgresBackupPage.ts:25]: localize

File: extensions/arc/src/ui/dashboards/postgres/postgresComputeStoragePage.ts [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+	protected get id(): string {
+		return 'postgres-compute-storage';
+	}
+
+	protected get icon(): { dark: string; light: string; } {
+		return IconPathHelper.computeStorage;
+	}
+
+	protected get container(): azdata.Component {
+		return this.modelView.modelBuilder.text().withProperties<azdata.TextComponentProperties>({ value: 'Compute + storage' }).component();
```
Comment (Charles-Gagnon) [inline extensions/arc/src/ui/dashboards/postgres/postgresComputeStoragePage.ts:25]: localize

File: extensions/arc/src/ui/dashboards/postgres/postgresDashboard.ts [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+		const computeStoragePage = new PostgresComputeStoragePage(modelView, this._controllerModel, this._databaseModel);
+		const connectionStringsPage = new PostgresConnectionStringsPage(modelView, this._controllerModel, this._databaseModel);
+		const backupPage = new PostgresBackupPage(modelView, this._controllerModel, this._databaseModel);
+		const propertiesPage = new PostgresPropertiesPage(modelView, this._controllerModel, this._databaseModel);
+		const networkingPage = new PostgresNetworkingPage(modelView, this._controllerModel, this._databaseModel);
+
+		return [
+			overviewPage.tab,
+			{
+				title: 'Settings',
```
Comment (Charles-Gagnon) [inline extensions/arc/src/ui/dashboards/postgres/postgresDashboard.ts:35]: localize this + Security below

File: extensions/arc/src/ui/dashboards/postgres/postgresDashboardPage.ts [hunk 13]
```diff
@@ -0,0 +1,15 @@
...
+ *  Copyright (c) Microsoft Corporation. All rights reserved.
+ *  Licensed under the Source EULA. See License.txt in the project root for license information.
+ *--------------------------------------------------------------------------------------------*/
+
+import * as azdata from 'azdata';
+import { ControllerModel } from '../../../models/controllerModel';
+import { PostgresModel } from '../../../models/postgresModel';
+import { DashboardPage } from '../../components/dashboardPage';
+
+export abstract class PostgresDashboardPage extends DashboardPage {
```
Comment (Charles-Gagnon) [inline extensions/arc/src/ui/dashboards/postgres/postgresDashboardPage.ts:11]: This class isn't really doing anything - unless you plan on adding something  why not just have the specific dashboard pages extend DashboardPage directly?

File: extensions/arc/src/ui/dashboards/postgres/postgresNetworkingPage.ts [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+	protected get id(): string {
+		return 'postgres-networking';
+	}
+
+	protected get icon(): { dark: string; light: string; } {
+		return IconPathHelper.networking;
+	}
+
+	protected get container(): azdata.Component {
+		return this.modelView.modelBuilder.text().withProperties<azdata.TextComponentProperties>({ value: 'Networking' }).component();
```
Comment (Charles-Gagnon) [inline extensions/arc/src/ui/dashboards/postgres/postgresNetworkingPage.ts:25]: localize

File: extensions/arc/src/ui/dashboards/postgres/postgresOverviewPage.ts [hunk 15]
```diff
@@ -0,0 +1,225 @@
...
+		}).component();
+
+		resetPasswordButton.onDidClick(async () => {
+			const password = await vscode.window.showInputBox({ prompt: loc.newPassword, password: true });
+			if (password === undefined) { return; }
+			await this.databaseModel.update(s => {
+				s.arc = s.arc ?? new DuskyObjectModelsDatabaseServiceArcPayload();
+				s.arc.servicePassword = password;
+			})
+				.then(_ => vscode.window.showInformationMessage(loc.passwordReset(this.databaseModel.fullName())))
```
Comment (Charles-Gagnon) [inline extensions/arc/src/ui/dashboards/postgres/postgresOverviewPage.ts:174]: Since these functions are async you can just await the calls each separately (and then wrap in try/catch to handle errors)

File: extensions/arc/src/ui/dashboards/postgres/postgresPropertiesPage.ts [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+
+		const endpoint: { ip?: string, port?: number } = this.databaseModel.endpoint();
+		const connectionString = `postgresql://postgres:${this.databaseModel.password()}@${endpoint.ip}:${endpoint.port}`;
+		const registration = this.controllerModel.registration('postgresInstances', this.databaseModel.namespace(), this.databaseModel.name());
+
+		const pairs: KeyValue[] = [
+			new InputKeyValue(loc.coordinatorEndpoint, connectionString),
+			new InputKeyValue(loc.postgresAdminUsername, 'postgres'),
+			new TextKeyValue(loc.status, this.databaseModel.service().status?.state ?? 'Unknown'),
+			new LinkKeyValue(loc.dataController, this.controllerModel.namespace(), _ => vscode.window.showInformationMessage('goto data controller')),
```
Comment (Charles-Gagnon) [inline extensions/arc/src/ui/dashboards/postgres/postgresPropertiesPage.ts:45]: Is the message a placeholder?

File: src/sql/azdata.proposed.d.ts [hunk 17]
```diff
@@ -306,6 +306,7 @@ declare module 'azdata' {
 
 	export interface InputBoxProperties extends ComponentProperties {
 		validationErrorMessage?: string;
+		readOnly?: boolean;
```
Comment (Charles-Gagnon) [inline src/sql/azdata.proposed.d.ts:309]: Let's do these ModelView changes in a different PR

File: extensions/arc/package.json [hunk 18]
```diff
@@ -0,0 +1,41 @@
...
+  },
+  "activationEvents": [
+    "*"
```
Comment (Charles-Gagnon) [inline extensions/arc/package.json:15]: Change this to 
```
"onCommand:arc.managePostgres",
"onCommand:arc.manageMiaa"
```

File: extensions/arc/package.json [hunk 19]
```diff
@@ -0,0 +1,41 @@
...
+  ],
+  "repository": {
+    "type": "git",
+    "url": "https://github.com/Microsoft/azuredatastudio.git"
+  },
+  "main": "./out/extension",
+  "contributes": {
+    "commands": [
+      {
+        "command": "arc.manageMiaa",
```
Comment (Charles-Gagnon) [inline extensions/arc/package.json:25]: Disable these in the command palette for merging into master

e.g. 
https://github.com/microsoft/azuredatastudio/blob/master/extensions/big-data-cluster/package.json#L41

File: extensions/arc/src/models/postgresModel.ts (hunk 8 above)
Comment (bergeron) [reply extensions/arc/src/models/postgresModel.ts:150]: Interesting, how should an amount of memory like '1GB' be localized?  It comes from k8s as '1G' and this function adds the 'B'.

File: extensions/arc/src/constants.ts [hunk 20]
```diff
@@ -0,0 +1,98 @@
...
+			dark: context.asAbsolutePath('images/security.svg')
+		};
+	}
+}
+
+export namespace cssStyles {
+	export const text = { 'user-select': 'text', 'cursor': 'text' };
+	export const title = { ...text, 'font-weight': 'bold', 'font-size': '14px' };
+	export const tableHeader = { ...text, 'text-align': 'left', 'border': 'none' };
+	export const tableRow = { ...text, 'border-top': 'solid 1px #ccc', 'border-bottom': 'solid 1px #ccc', 'border-left': 'none', 'border-right': 'none' };
```
Comment (alanrenmsft) [inline extensions/arc/src/constants.ts:96]: is color #ccc good for all themes?: light, dark, high contrast

File: extensions/arc/src/ui/dashboards/postgres/postgresDashboardPage.ts (hunk 13 above)
Comment (bergeron) [reply extensions/arc/src/ui/dashboards/postgres/postgresDashboardPage.ts:11]: It gives all the Postgres pages access to the PostgresModel and ControllerModel.  Rather than define this constructor on all the Postgres pages.

File: extensions/arc/src/ui/dashboards/postgres/postgresOverviewPage.ts (hunk 15 above)
Comment (bergeron) [reply extensions/arc/src/ui/dashboards/postgres/postgresOverviewPage.ts:174]: Is the suggestion to not use .then/.catch and instead do:

```
try {
    await this.databaseModel.createDatabase(db);
    vscode.window.showInformationMessage(...);
} catch (error) {
    vscode.window.showErrorMessage(...);
}
```

File: extensions/arc/src/ui/dashboards/postgres/postgresPropertiesPage.ts (hunk 16 above)
Comment (bergeron) [reply extensions/arc/src/ui/dashboards/postgres/postgresPropertiesPage.ts:45]: Oh yeah the mockups have these as hyperlinks.  I assume they're not https:// links but instead it sends you to the manage controller page inside ADS.

File: extensions/arc/src/constants.ts (hunk 20 above)
Comment (bergeron) [reply extensions/arc/src/constants.ts:96]: Good question.  I'll have to test all the themes, but I copied this style from existing tables in ADS.

File: extensions/arc/src/ui/dashboards/postgres/postgresOverviewPage.ts (hunk 15 above)
Comment (Charles-Gagnon) [reply extensions/arc/src/ui/dashboards/postgres/postgresOverviewPage.ts:174]: Yeah - slightly cleaner and easier to follow

File: extensions/arc/src/constants.ts (hunk 20 above)
Comment (bergeron) [reply extensions/arc/src/constants.ts:96]: It looks good to me in all the themes, though I don't know what the accessibility standard is.

File: extensions/arc/src/models/postgresModel.ts (hunk 8 above)
Comment (bergeron) [reply extensions/arc/src/models/postgresModel.ts:150]: I assume this means the unit (B, KB, KiB, etc) should be localized.  I can add localized constants for those.

Comment (Charles-Gagnon) [reply extensions/arc/src/models/postgresModel.ts:150]: Well that's weird. Ok for now let's just leave these as is until we decide it's worth spending time looking into how best to handle this.

File: extensions/arc/src/ui/components/keyValueContainer.ts [hunk 21]
```diff
@@ -0,0 +1,95 @@
...
+			value: this.value // TODO: Add a readOnly property to input boxes
+		}).component());
+
+		const copy = modelBuilder.button().withProperties<azdata.ButtonProperties>({
+			iconPath: IconPathHelper.copy, width: '17px', height: '17px'
+		}).component();
+
+		copy.onDidClick(async () => {
+			vscode.env.clipboard.writeText(this.value);
+			vscode.window.showInformationMessage(`${this.key} copied to clipboard`);
```
Comment (Charles-Gagnon) [inline extensions/arc/src/ui/components/keyValueContainer.ts:73]: Localize

File: extensions/arc/src/localizedConstants.ts [hunk 22]
```diff
@@ -46,6 +46,7 @@ export const feedback = localize('arc.feedback', 'Feedback');
 export const selectConnectionString = localize('arc.selectConnectionString', 'Select from available client connection strings below');
 export const vCores = localize('arc.vCores', 'vCores');
 export const ram = localize('arc.ram', 'RAM');
+export const copiedToClipboard = localize('arc.copiedToClipboard', 'copied to clipboard');
```
Comment (Charles-Gagnon) [inline extensions/arc/src/localizedConstants.ts:49]: Use a parameter for this for easier translation into non Left-To-Right languages

**Comment by bergeron** [comment]
> Looks like you have a conflict in gulp.hygiene.js now actually

The build is failing because of a merge conflict in that file.  But I'm confused because I can't reproduce any conflict.  This PR doesn't report a conflict, and I don't get a conflict when I merge my branch into master on the command line.

File: build/gulpfile.hygiene.js [hunk 23]
```diff
@@ -109,7 +109,8 @@ const indentationFilter = [
 	'!extensions/sql-database-projects/src/test/baselines/*.sqlproj',
 	'!extensions/big-data-cluster/src/bigDataCluster/controller/apiGenerated.ts',
 	'!extensions/big-data-cluster/src/bigDataCluster/controller/clusterApiGenerated2.ts',
-	'!resources/linux/snap/electron-launch'
+	'!resources/linux/snap/electron-launch',
```
Comment (aaomidi) [inline build/gulpfile.hygiene.js:112]: You can't add stuff here to the bottom of the list, you need to add it to the middle of the list or you'll get conflicts like this.

File: extensions/arc/src/controller/generated/dusky/api/databaseRouterApi.ts [hunk 24]
```diff
@@ -0,0 +1,2647 @@
...
+            authenticationPromise = authenticationPromise.then(() => this.authentications.BearerAuth.applyToRequest(localVarRequestOptions));
+        }
+        authenticationPromise = authenticationPromise.then(() => this.authentications.default.applyToRequest(localVarRequestOptions));
+
+        let interceptorPromise = authenticationPromise;
+        for (const interceptor of this.interceptors) {
+            interceptorPromise = interceptorPromise.then(() => interceptor(localVarRequestOptions));
+        }
+
+        return interceptorPromise.then(() => {
```
Comment (aaomidi) [inline extensions/arc/src/controller/generated/dusky/api/databaseRouterApi.ts:174]: You're in async/await, why do this?

File: extensions/arc/src/controller/generated/dusky/api/databaseRouterApi.ts [hunk 25]
```diff
@@ -0,0 +1,2647 @@
...
+            method: 'POST',
+            qs: localVarQueryParameters,
+            headers: localVarHeaderParams,
+            uri: localVarPath,
+            useQuerystring: this._useQuerystring,
+            json: true,
+            body: ObjectSerializer.serialize(duskyObjectModelsDatabase, "DuskyObjectModelsDatabase")
+        };
+
+        let authenticationPromise = Promise.resolve();
```
Comment (aaomidi) [inline extensions/arc/src/controller/generated/dusky/api/databaseRouterApi.ts:160]: This code style is also confusing

File: extensions/arc/src/controller/generated/dusky/api/databaseRouterApi.ts (hunk 24 above)
Comment (aaomidi) [reply extensions/arc/src/controller/generated/dusky/api/databaseRouterApi.ts:174]: FYI I'm not blocking on this, just noticed it.

File: extensions/arc/src/controller/generated/dusky/api/databaseRouterApi.ts (hunk 25 above)
Comment (aaomidi) [reply extensions/arc/src/controller/generated/dusky/api/databaseRouterApi.ts:160]: FYI I'm not blocking on this, just noticed it.

File: extensions/arc/src/controller/generated/dusky/api/databaseRouterApi.ts (hunk 24 above)
Comment (Charles-Gagnon) [reply extensions/arc/src/controller/generated/dusky/api/databaseRouterApi.ts:174]: Yeah so this is all generated code so isn't stuff we wrote (although if we do see any actual bugs we should fix them for sure)

File: build/gulpfile.hygiene.js (hunk 23 above)
Comment (Charles-Gagnon) [reply build/gulpfile.hygiene.js:112]: Well that's weird. Why? Is the distro merge adding stuff onto the end that it's conflicting with or something?

Comment (aaomidi) [reply build/gulpfile.hygiene.js:112]: Yeah the distro has stuff at the end of this. There's a cleanup we can do there to make sure this stops conflicting.

@anthonydresser do you think we can make a new array in distro and append it programmatically to this array? I've seen this issue come up a bunch of times now.

File: extensions/arc/src/controller/generated/dusky/api/databaseRouterApi.ts (hunk 25 above)
Comment (bergeron) [reply extensions/arc/src/controller/generated/dusky/api/databaseRouterApi.ts:160]: Yeah the code generated by openapi-generator isn't perfect.  But it's very convenient to generate all the routes and models from the controller's swagger spec.  openapi-generator was chosen over editor.swagger.io because editor.swagger.io doesn't support OpenAPI 3.x.

File: extensions/arc/src/ui/dashboards/postgres/postgresDashboard.ts [hunk 26]
```diff
@@ -0,0 +1,51 @@
...
+import { PostgresNetworkingPage } from './postgresNetworkingPage';
+import { Dashboard } from '../../components/dashboard';
+
+export class PostgresDashboard extends Dashboard {
+	constructor(title: string, private _controllerModel: ControllerModel, private _databaseModel: PostgresModel) {
+		super(title);
+	}
+
+	protected async registerTabs(modelView: azdata.ModelView): Promise<(azdata.DashboardTab | azdata.DashboardTabGroup)[]> {
+		await Promise.all([this._controllerModel.refresh(), this._databaseModel.refresh()]);
```
Comment (Charles-Gagnon) [inline extensions/arc/src/ui/dashboards/postgres/postgresDashboard.ts:24]: We'll want to change this before we ship so that there's some indication that it's in a loading state - otherwise it might just be sitting there on an empty blank page for a while if the controller is slow to respond.
